File: downloading.py
from configparser import ConfigParser
import json
import os
from typing import List
import requests
import os
import dotenv
dotenv.load_dotenv()
USER = os.getenv('GITHUB_USER')
TOKEN = os.getenv('GITHUB_PAT')


def query_repos(query:str,page:int,config:dict) -> List[dict]:
    num_repos = int(config['repos']['num_repos'])
    languages = config['repos']['languages']
    if num_repos<page+1*100:
        per_page = num_repos%100
    else:
        per_page = 100
    print(f'Querying page {page} with {per_page} results')
    languages_str = ''
    for lang in languages:
        print(lang)
        languages_str+=f'language:{lang}+'
    query_str = f"{query} {languages_str} &per_page={per_page}&page={page}&sort=stars"
    query_str = query_str.replace(" ", "+")
    query_url = f'https://api.github.com/search/repositories?q={query_str}'
    r = requests.get(
            query_url,
            auth = (USER, TOKEN)
            )
    data = r.json()
    items = []
    for idx,x in enumerate(data['items']):
        if idx+1 > num_repos:
            break
        items.append(x)
    return items

def get_repos(query:str,config:ConfigParser):
    print('Getting repos')
    file_directory = './data/'+query.replace(" ", "_")
    if not os.path.exists(file_directory):
        os.mkdir(file_directory)
    query_results = []
    for page_num in range(5):
        query_results.extend(query_repos(query,page_num,config))
        if page_num+1*100 > config['repos']['num_repos']:
            break
    with open(file_directory+'/github_repos.json', 'w') as f:
        json.dump(query_results, f)

def download_repos(query:str):
    file_directory = './data/'+query.replace(" ", "_")
    repos_file = file_directory+'/github_repos.json'
    download_dir = file_directory+'/repos'
    if not os.path.exists(download_dir):
        os.mkdir(download_dir)
    with open(repos_file,'r') as f:
        data = json.load(f)

    for x in data:
        print(x['name'], x['stargazers_count'])
        # git.Git("./repos/"+x['name']).clone(x['clone_url'])
        os.system('git clone '+x['clone_url']+f' {download_dir}/'+x['name'])

File: src/downloading/downloading.ts
import * as fs from "fs";
import { join } from "path";
import * as dotenv from 'dotenv'
import axios from 'axios'
import { exec } from "child_process";

dotenv.config();

const USER = process.env.GITHUB_USER;
const TOKEN = process.env.GITHUB_PAT;

async function queryRepos(
  query: string,
  page: number,
  config: any
): Promise<any[]> {
  const numRepos = config.repos.numRepos;
  const languages = config.repos.languages;
  const perPage =
    numRepos < page + 1 * 100 ? numRepos % 100 : 100;
  console.log(`Querying page ${page} with ${perPage} results`);
  let languagesStr = "";
  for (const lang of languages) {
    console.log(lang);
    languagesStr += `language:${lang}+`;
  }
  const queryStr = `${query} ${languagesStr} &per_page=${perPage}&page=${page}&sort=stars`;
  const queryUrl = `https://api.github.com/search/repositories?q=${queryStr}`;
  const headers = {
    Authorization: `Bearer ${TOKEN}`,
  }
  const req = await axios.get(queryUrl, {
    headers: headers
  })
  const data = req.data;
  const items: any[] = [];
  for (let idx = 0; idx < data.items.length; idx++) {
    if (idx + 1 > numRepos) {
      break;
    }
    items.push(data.items[idx]);
  }
  return items;
}

export async function getRepos(query: string, config:any): Promise<void> {
  console.log("Getting repos");
  const fileDirectory = join("./data", query.replace(" ", "_"));
  if (!(await fs.existsSync(fileDirectory))) {
    await fs.mkdirSync(fileDirectory);
  }
  const queryResults: any[] = [];
  for (let pageNum = 0; pageNum < 5; pageNum++) {
    queryResults.push(...(await queryRepos(query, pageNum, config)));
    if (pageNum + 1 * 100 > config.repos.numRepos) {
      break;
    }
  }
  await fs.writeFileSync(
    join(fileDirectory, "github_repos.json"),
    JSON.stringify(queryResults)
  );
}

export async function downloadRepos(query: string): Promise<void> {
  const fileDirectory = join("./data", query.replace(" ", "_"));
  const reposFile = join(fileDirectory, "github_repos.json");
  const downloadDir = join(fileDirectory, "repos");
  if (!(await fs.existsSync(downloadDir))) {
    await fs.mkdirSync(downloadDir);
  }
  const data: any[] = JSON.parse(
    await fs.readFileSync(reposFile, { encoding: "utf-8" })
  );
  for (const x of data) {
    console.log(x.name, x.stargazers_count);
    await exec(`git clone ${x.clone_url} ${join(downloadDir, x.name)}`);
  }
}

export async function downloadRepo(repo: string): Promise<void> {
  return new Promise((resolve, reject) => { 
    const downloadDir = "/tmp/data/repos"
  if (!(fs.existsSync('/tmp/data'))) {
      fs.mkdirSync('/tmp/data');
    }
  if (!(fs.existsSync(downloadDir))) {
    fs.mkdirSync(downloadDir);
  }

  const proc = exec(`git clone https://github.com/${repo}.git ${join(downloadDir, repo.split("/")[1])}`);
    proc.on('exit', (code) => { 
      console.log(`Child exited with code ${code}`);
      resolve();
  })
})
}